presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/GeometryJtsUtils.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.geospatial;
+
+import io.airlift.slice.BasicSliceInput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import org.locationtech.jts.geom.Coordinate;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.GeometryFactory;
+import org.locationtech.jts.geom.LineString;
+import org.locationtech.jts.geom.LinearRing;
+import org.locationtech.jts.geom.Point;
+import org.locationtech.jts.geom.Polygon;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.facebook.presto.geospatial.GeometryUtils.isEsriNaN;
+import static io.airlift.slice.SizeOf.SIZE_OF_DOUBLE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+
+public class GeometryJtsUtils
+{
+    private static final int POINT = 1;
+    private static final int POLYLINE = 3;
+    private static final int POLYGON = 5;
+    private static final int MULTIPOINT = 8;
+    private static final GeometryFactory FACTORY = new GeometryFactory();
+
+    private GeometryJtsUtils() {}
+
+    /**
+     * Deserializes ESRI shape as described in
+     * https://www.esri.com/library/whitepapers/pdfs/shapefile.pdf
+     * into JTS Geometry object.
+     */
+    public static Geometry deserialize(Slice shape)
+    {
+        if (shape == null) {
+            return null;
+        }
+
+        BasicSliceInput input = shape.getInput();
+        input.skip(SIZE_OF_INT); // SRID
+
+        // GeometryCollection: spatialReferenceId|len-of-shape1|bytes-of-shape1|len-of-shape2|bytes-of-shape2...
+        List<Geometry> geometries = new ArrayList<>();
+        while (input.available() > 0) {
+            int length = input.readInt();
+            geometries.add(readGeometry(input.readSlice(length).getInput()));
+        }
+
+        if (geometries.isEmpty()) {
+            return FACTORY.createGeometryCollection();
+        }
+        else if (geometries.size() == 1) {
+            return geometries.get(0);
+        }
+        return FACTORY.createGeometryCollection(geometries.toArray(new Geometry[0]));
+    }
+
+    private static Geometry readGeometry(SliceInput input)
+    {
+        int geometryType = input.readInt();
```

Comment:
`requireNonNull(input, "input is null")` (applies to other methods in this class)